File: Node.js/src/inOrder.ts
import { ArgType, compareSubfs, mainSubf as main } from './common';

/**
 * @description Test the internal alphabetical order of words. Written with branch prediction in mind.
 * 
 * @param {string} text - Words to test.
 * @returns {string} Newline-deliniated list of words and whethere or not they are in alphabetical order.
 */
export const inOrderBranchPrediction = (text: string): string => {
    const results: string[] = text.split(" ").map((word: string) => {
        const order: boolean = wordInOrderBranchPrediction(word);
        const result: string = `${word} -${order ? '' : ' not'} in order`;
        return result;
    });
    return results.join("\n");
};

/**
 * @description Test if the input word's characters are in alphabetical order. Written with branch prediction in mind.
 * 
 * @param {string} word - Word to test.
 * @returns {boolean} If the word is alphabetical or not.
 */
export const wordInOrderBranchPrediction = (word: string): boolean => {
    const characters = word.split("");
    for (let i = 0; i < characters.length - 1; i++) {
        if (characters[i] <= characters[i + 1]) {
            continue;
        }
        return false;
    }
    return true;
};

/**
 * @description Test the internal alphabetical order of words. Written with readability in mind.
 * 
 * @param {string} text - Words to test.
 * @returns {string} Newline-deliniated list of words and whethere or not they are in alphabetical order.
 */
export const inOrderReadability = (text: string): string => {
    const results: string[] = text.split(" ").map((word: string) => {
        const order: boolean = wordInOrderReadability(word);
        const result: string = `${word} -${order ? '' : ' not'} in order`;
        return result;
    });
    return results.join("\n");
};

/**
 * @description Test if the input word's characters are in alphabetical order. Written with readability in mind.
 * 
 * @param {string} word - Word to test.
 * @returns {boolean} If the word is alphabetical or not.
 */
export const wordInOrderReadability = (word: string): boolean => {
    const characters = word.split("");
    for (let i = 0; i < characters.length - 1; i++) {
        if (characters[i] > characters[i + 1]) {
            return false;
        }
    }
    return true;
};

/**
 * @description Test the internal alphabetical order of words. Uses a simple, probabability-based heuristic function on each word to guess the quicker `wordInOrder...` function to use.
 * 
 * @param {string} text - Words to test.
 * @returns {string} Newline-deliniated list of words and whethere or not they are in alphabetical order.
 */
export const inOrderHeuristic = (text: string): string => {
    const results: string[] = text.split(" ").map((word: string) => {
        const order: boolean = word.charAt(0) > 'm' ? wordInOrderReadability(word) : wordInOrderBranchPrediction(word);
        const result: string = `${word} -${order ? '' : ' not'} in order`;
        return result;
    });
    return results.join("\n");
}

main({
    branchPrediction: inOrderBranchPrediction,
    readability: inOrderReadability,
    heuristic: inOrderHeuristic,
    compare: compareSubfs
}, ArgType.STRING, (value: string) => value.length, "What word(s)?");

File: Go/inOrder.go
package main

import (
	"fmt"
	"strings"
)

func InOrder(text string, f func(word []rune) bool) string {
	words := strings.Split(strings.ToLower(text), " ")
	var results []string
	for _, word := range words {
		order := f([]rune(word))
		var verdict string
		if order {
			verdict = ""
		} else {
			verdict = "not "
		}
		results = append(results, fmt.Sprintf("%v - %vin order", word, verdict))
	}
	return strings.Join(results, "\n")
}

func InOrderBranchPrediction(text string) string {
	return InOrder(text, WordInOrderBranchPrediction)
}

func InOrderReadability(text string) string {
	return InOrder(text, WordInOrderReadability)
}

func InOrderHeuristic(text string) string {
	words := strings.Split(strings.ToLower(text), " ")
	var results []string
	for _, word := range words {
		var order bool
		if word[0] > 'm' {
			order = WordInOrderReadability([]rune(word))
		} else {
			order = WordInOrderBranchPrediction([]rune(word))
		}
		var verdict string
		if order {
			verdict = ""
		} else {
			verdict = "not "
		}
		results = append(results, fmt.Sprintf("%v - %vin order", word, verdict))
	}
	return strings.Join(results, "\n")
}

func WordInOrderBranchPrediction(word []rune) bool {
	for i, character := range word[:len(word)-1] {
		if character <= word[i+1] {
			continue
		}
		return false
	}
	return true
}

func WordInOrderReadability(word []rune) bool {
	for i, character := range word[:len(word)-1] {
		if character > word[i+1] {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(InOrderBranchPrediction("almost chef"))
	fmt.Println(InOrderReadability("bit dig"))
	fmt.Println(InOrderHeuristic("billowy nope"))
}
